[nativert] Move DelegateExecutor to PyTorch core

yiming0416 opened the pull request:
Summary:
Moves DelegateExecutor base class to PyTorch core. It provides the extension point of backend delegation for NativeRT.
Torch Native Runtime RFC: pytorch/rfcs#72

Test Plan:
This is only a virtual base class. So relying on internal CI is sufficient.

Rollback Plan:

Differential Revision: D76351984

zhxchen17 commented on torch/nativert/executor/DelegateExecutor.cpp:
```diff
@@ -0,0 +1,65 @@
+std::string extractToTemporaryFolder(
+    caffe2::serialize::PyTorchStreamReader& packageReader,
+    const std::string& targetPath) {
+  char outputDir[] = "/tmp/model_XXXXXX";
```
@yiming0416 seems the directory name here is a bit rough, could we do something better here? (e.g. use a more descriptive name + a random uuid)

zhxchen17 commented on torch/nativert/executor/DelegateExecutor.cpp:
```diff
@@ -0,0 +1,98 @@
+}
+
+char* generateOutputDir() {
+  static char outputDir[64];
```
this is not thread safe, please just use a std::string

zhxchen17 reviewed (changes requested):
@yiming0416 overall seems okay to me, just some thoughts on output dir.

yiming0416 committed:
[nativert] Move DelegateExecutor to PyTorch core (#155581)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/155581

Moves DelegateExecutor base class to PyTorch core. It provides the extension point of backend delegation for NativeRT.
Torch Native Runtime RFC: pytorch/rfcs#72

Test Plan:
This is only a virtual base class. So relying on internal CI is sufficient.

Rollback Plan:

Reviewed By: zhxchen17

Differential Revision: D76351984

yiming0416 commented:
@pytorchbot merge -f "landed internally"
